### PR TITLE
#HK-3 Develop players absence tracking

### PR DESCRIPTION
**Pull request by AlexeyFilich**
Solves #3. Implement check for all players absence on plugin start

### CODE REVIEW

File: src/main/java/me/drendov/HouseKeeping/HouseKeeping.java [hunk 1]
```diff
@@ -191,4 +194,37 @@ synchronized public String getMessage(Messages messageID, String... args) {
...
+            getPluginLoader().disablePlugin(this);
+            return;
+        }
+    }
+
+
+    private void checkAbsence() {
+        Essentials ess = (Essentials)Bukkit.getPluginManager().getPlugin("Essentials");
+        File csv = new File(ess.getDataFolder(), "usermap.csv");
+        UserMap userMap = ess.getUserMap();
```
Comment (DmitryRendov) [inline src/main/java/me/drendov/HouseKeeping/HouseKeeping.java:215]: I don't think that we can count on this file.
Have you had a look at getOfflinePlayers() or getOnlinePlayers()?

**Comment by DmitryRendov** [comment]
Also, this ess metthod might help https://github.com/EssentialsX/Essentials/blob/555a62c58203dd310ffb32bf08bbd28c6709401d/Essentials/src/main/java/com/earth2me/essentials/BalanceTopImpl.java#L31

File: src/main/java/me/drendov/HouseKeeping/HouseKeeping.java (hunk 1 above)
Comment (AlexeyFilich) [reply src/main/java/me/drendov/HouseKeeping/HouseKeeping.java:215]: `Bukkit.getOfflinePlayers()` does not report any players listed in `plugins/Essentials/playerdata`. Essentials does not seem to provide any cleaner way to list all players. `UserMap` should provide a method to get all uniuqe players, but it seems to return empty set all the time during testing.

File: .vscode/launch.json [hunk 2]
```diff
@@ -2,11 +2,12 @@
     "version": "0.2.0",
     "configurations": [
         {
-            "type": "java",
             "name": "Debug plugin",
+            "type": "java",
```
Comment (DmitryRendov) [inline .vscode/launch.json]: Please, remove from the repo all IDE specific settings, please

File: .vscode/tasks.json [hunk 3]
```diff
@@ -1,15 +1,58 @@
 {
     "version": "2.0.0",
     "tasks": [
+        {
```
Comment (DmitryRendov) [inline .vscode/tasks.json]: the same as above

File: src/main/java/me/drendov/HouseKeeping/HouseKeeping.java [hunk 4]
```diff
@@ -67,11 +69,14 @@ public void onEnable() {
...
             Metrics metrics = new Metrics(this, pluginId);
-        }
-        catch (Throwable ignored){}
+        } catch (Throwable ignored) {}
         PluginDescriptionFile pdfFile = this.getDescription();
         this.version = pdfFile.getVersion();
         logger.info("HouseKeeping v" + this.version + " enabled.");
+
+        // Schedule absence check in 5 seconds
+        ScheduledExecutorService runAbsenceCheck = Executors.newSingleThreadScheduledExecutor();
```
Comment (DmitryRendov) [inline src/main/java/me/drendov/HouseKeeping/HouseKeeping.java]: Could you please do it configurable, eg if "processAtStart = true", then list players
Default is true.

File: src/main/java/me/drendov/HouseKeeping/HouseKeeping.java [hunk 5]
```diff
@@ -190,5 +197,27 @@ synchronized public String getMessage(Messages messageID, String... args) {
...
+
+        Date date = new Date();
+
+        long now = date.getTime();
+        long minBeforeAbsent = now - (long)HouseKeeping.getInstance().config.getDaysWithholdFunds() * (long)86400000;
 
+        for (UUID uuid : ess.getUserMap().getAllUniqueUsers()) {
+            User player = userMap.getUser(uuid);
+            if (player != null && player.getLastLogin() < minBeforeAbsent)
+                logger.info(String.format("Player %s is absent for %d days (%s$)", player.getName(), (now - player.getLastLogin()) / 86400000, player.getMoney().toString()));
```
Comment (DmitryRendov) [inline src/main/java/me/drendov/HouseKeeping/HouseKeeping.java]: Please use short version, eg "%s, % days, $%s"
Also, round up a balance to whole numbers

**Review by DmitryRendov — DISMISSED** [review]

**Review by DmitryRendov — APPROVED** [review]

**Pull request merged**